src/render.rs

```diff
@@ -420,15 +492,165 @@ impl RenderData {
                 }
             }
 
-            if note_keys_len >= 2 {
+            if note_keys_len >= 2 && !self.needs_note() && !self.needs_profile() {
                 break;
             }
+        }
+
+        match fetch_handle.await {
+            Ok(Ok(())) => Ok(()),
+            Ok(Err(err)) => Err(err),
+            Err(join_err) => Err(Error::Generic(format!(
+                "relay fetch task failed: {}",
+                join_err
+            ))),
+        }
+    }
+}
 
-            loops += 1;
+fn collect_relay_hints(event: &Event) -> Vec<RelayUrl> {
+    let mut relays = Vec::new();
+    for tag in event.tags.as_slice() {
+        let parts = tag.as_slice();
+        if parts.is_empty() {
+            continue;
         }
+        let tag_name = parts[0].as_str();
+        let candidate = if matches!(tag_name, "r" | "relay" | "relays") {
```

Comment:
this is confusing to me. as_str should be an Option<&str> yet the matches! doesn't seem to match on an Option?

this is all slop, we should write code normally here and pull out a string with parts.get_str(0)